### ISSUE TITLE
test_sandbox_varargs and test_sandbox_va_copy fail for 128-bit capabilities

### ISSUE THREAD

**Issue by staceyson** [CTSRD-CHERI/clang#89]
Using QEMU-CHERI compiled with '-DCHERI_MAGIC128' (magic 128-bit capabilities) running CheriBSD (CHERI128_MALTA64 kern config) the following two cheritests are failing:

TEST: test_sandbox_varargs: Verify that varargs work in a sandbox
FAIL: test_sandbox_varargs: Unexpected trusted stack unwind
TEST: test_sandbox_va_copy: Verify that va_copy works in a sandbox
FAIL: test_sandbox_va_copy: Unexpected trusted stack unwind

And for cheriabitest they both also fail (in two different ways):

TEST: test_sandbox_varargs: (null)
FAIL: (null): test_sandbox_varargs
TEST: test_sandbox_va_copy: (null)
FAIL: (null): test_sandbox_va_copy

**Comment by davidchisnall** [comment]
Do the CHERI C test suite variadic tests pass?

**Comment by staceyson** [comment]
Unfortunately I am not having any luck with cheri-c-tests:

(1) They don't build in CheriBSD:
```
===> bin/cheriabitest (all)
/home/sson/src/llvm/Build/bin/clang -g -integrated-as --target=cheri-unknown-freebsd  -msoft-float --sysroot=/var/tmp/obj/mips.mips64/home/sson/src/cheribsd/tmp -mabi=sandbox -mxgot -cheri-linker -mllvm -cheri128 -O -pipe -DCHERI_C_TESTS  -I/home/sson/src/cheribsd/contrib/cheri-c-tests -DTEST_CUSTOM_FRAMEWORK -I/home/sson/src/cheribsd/bin/cheriabitest/../cheritest -G0   -O0 -Wno-error -Qunused-arguments -g -MD -MP -MF.depend.test_runtime.o -MTtest_runtime.o -std=gnu99 -Wsystem-headers -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wshadow -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wnested-externs -Wredundant-decls -Wold-style-definition -Wno-pointer-sign -Wmissing-variable-declarations -Wthread-safety -Wno-empty-body -Wno-string-plus-int -Wno-unused-const-variable  -Qunused-arguments -c /home/sson/src/cheribsd/contrib/cheri-c-tests/test_runtime.c -o test_runtime.o
In file included from /home/sson/src/cheribsd/contrib/cheri-c-tests/test_runtime.c:40:
In file included from /home/sson/src/cheribsd/contrib/cheri-c-tests/cheri_c_test.h:64:
/home/sson/src/cheribsd/bin/cheriabitest/../cheritest/cheri_c_test_framework.h:7:9: warning: 
      'assert' macro redefined [-Wmacro-redefined]
#define assert(x)                                                       \
        ^
/var/tmp/obj/mips.mips64/home/sson/src/cheribsd/tmp/usr/include/assert.h:54:9: note: 
      previous definition is here
#define assert(e)       ((e) ? (void)0 : __assert(__func__, __FILE__, \
        ^
/home/sson/src/cheribsd/contrib/cheri-c-tests/test_runtime.c:43:27: warning: unused
      parameter 'capreg' [-Wunused-parameter]
static void handler(void *capreg, int cause)
                          ^
/home/sson/src/cheribsd/contrib/cheri-c-tests/test_runtime.c:43:39: warning: unused
      parameter 'cause' [-Wunused-parameter]
static void handler(void *capreg, int cause)
                                      ^
/home/sson/src/cheribsd/contrib/cheri-c-tests/test_runtime.c:159:2: error: static_assert
      failed "Layout of struct cheri_frame has changed!"
        _Static_assert(offsetof(struct cheri_frame, cf_pcc) == 27*32,
        ^              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/sson/src/cheribsd/contrib/cheri-c-tests/test_runtime.c:171:12: warning: comparison of
      integers of different signs: '__attribute__((address_space(200))) register_t' (aka
      '__attribute__((address_space(200))) long') and 'unsigned long' [-Wsign-compare]
        assert(pc < __builtin_memcap_length_get(getCHERIFrame(context)->cf_pcc));
               ~~ ^                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/sson/src/cheribsd/bin/cheriabitest/../cheritest/cheri_c_test_framework.h:9:8: note: 
      expanded from macro 'assert'
        if (!(x))                                                       \
              ^
/home/sson/src/cheribsd/contrib/cheri-c-tests/test_runtime.c:354:37: warning: unused
      parameter 'info' [-Wunused-parameter]
capsighandler(int signo, siginfo_t *info, ucontext_t *uap)
                                    ^
/home/sson/src/cheribsd/contrib/cheri-c-tests/test_runtime.c:385:6: warning: no previous
      prototype for function 'test_setup' [-Wmissing-prototypes]
void test_setup(void)
     ^
6 warnings and 1 error generated.
*** Error code 1
```

(2) My 128-bit cross built  cheri-c-tests all crash with a length violation:
```
root@cheri:/var/tmp/cheri-c-tests # ./va_args
CHERI cause: ExcCode: 0x01 RegNum: $c02 (length violation)
$c00: v:1 s:0 p:0007807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c01: v:1 s:0 p:0007807d b:0000007fffffe880 l:0000000000000020 o:0 t:0
$c02: v:1 s:0 p:0007807d b:0000007fffffe8e0 l:0000000000000060 o:0 t:0
$c03: v:1 s:0 p:0007807d b:0000007fffffe7e0 l:0000000000000020 o:0 t:0
$c04: v:1 s:0 p:0007807d b:0000007fffffe5e0 l:0000000000000020 o:0 t:0
$c05: v:1 s:0 p:0007807d b:0000000000000000 l:0000010000000000 o:1200bbd20 t:0
$c06: v:1 s:0 p:0007807d b:0000007fffffe5a0 l:0000000000000020 o:0 t:0
$c07: v:1 s:0 p:0007807d b:0000000000000000 l:0000010000000000 o:12009be88 t:0
$c08: v:1 s:0 p:0007807d b:0000007fffffe7e0 l:0000000000000020 o:0 t:0
$c09: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c10: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c11: v:1 s:0 p:0007807d b:0000007fff7ff000 l:00000000007ff8e0 o:0 t:0
$c12: v:1 s:0 p:0007807d b:0000007fffffe640 l:0000000000000020 o:0 t:0
$c13: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c14: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c15: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c16: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c17: v:1 s:0 p:00078017 b:0000000000000000 l:0000010000000000 o:120002360 t:0
$c18: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c19: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c20: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c21: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c22: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c23: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c24: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c26: v:1 s:0 p:0007807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c31: v:1 s:0 p:00078017 b:0000000000000000 l:0000010000000000 o:12000239c t:0
Feb 20 21:37:53 cheri kernel: USER_CHERI_EXCEPTION: pid 627 tid 100040 (va_args), uid 0: CP2 fault (type 0x32)
Feb 20 21:37:53 cheri kernel: Trapframe Register Dump:
Feb 20 21:37:53 cheri kernel: zero: 0   at: 0x80        v0: 0x1200ca790 v1: 0x1
Feb 20 21:37:53 cheri kernel: a0: 0x7ff698      a1: 0   a2: 0   a3: 0
Feb 20 21:37:53 cheri kernel: a4: 0     a5: 0   a6: 0   a6: 0
Feb 20 21:37:53 cheri kernel: t0: 0     t1: 0   t2: 0   t3: 0
Feb 20 21:37:53 cheri kernel: t8: 0     t9: 0x1200b2000 s0: 0   s1: 0
Feb 20 21:37:53 cheri kernel: s2: 0     s3: 0   s4: 0   s5: 0
Feb 20 21:37:53 cheri kernel: s6: 0     s7: 0   k0: 0   k1: 0
Feb 20 21:37:53 cheri kernel: gp: 0     sp: 0x7ff6a0    s8: 0x7ff6a0    ra: 0
Feb 20 21:37:53 cheri kernel: sr: 0x408084b3    mullo: 0        mulhi: 0        badvaddr: 0x12000239c
Feb 20 21:37:53 cheri kernel: cause: 0x48       pc: 0x12000239c
Signal 34 (core dumped)
```

Please note that I had to remove "-cheri-linker" from TEST_CFLAGS to cross build.  Otherwise:
```
 /home/sson/src/llvm/Build128/bin/clang -c -g -mabi=sandbox -cheri-linker -Werror -O3 test_runtime.c -o test_runtime.o
clang-3.8: error: argument unused during compilation: '-cheri-linker'
Makefile:58: recipe for target 'test_runtime.o' failed
```
However, I do have that problem cross building for 256-bit capabilities.

**Comment by davidchisnall** [comment]
The unused argument thing is a clang bug.  Turning off `-Werror` should be enough to work around it.

The `_Static_assert` that was failing in the build should be fixed now.  The fact that is *wasn't* failing for you in cross building is slightly concerning, as it implies that you're building for 32-byte capabilities.  Unfortunately, we use the same ELF machine type for both, so it's hard to tell if you're accidentally running the wrong kind of binary.  You need to add `-mllvm -cheri128` to your `CFLAGS` when compiling for CHERI128.

**Comment by staceyson** [comment]
OK, the cheri-c-tests integrated with cheriabitest now builds (in CheriBSD) will the latest cheri-c-tests commit.  Here are the tests that are failing:

```
sson@cheri:/var/tmp % ./cheriabitest -a
[...]
TEST: cheritest_vm_swap: (null)
X(null): FAIL: (null) (cheritest_vm_swap)
[...]
TEST: test_sandbox_divzero_catch: (null)
FAIL: (null): test_sandbox_divzero_catch
TEST: test_sandbox_divzero_nocatch: (null)
FAIL: (null): test_sandbox_divzero_nocatch
[...]
TEST: test_sandbox_varargs: (null)
FAIL: (null): test_sandbox_varargs
TEST: test_sandbox_va_copy: (null)
FAIL: (null): test_sandbox_va_copy
[...]
TEST: cheri_c_test_intcap: (null)
FAIL: (null): cheri_c_test_intcap
[...]
TEST: cheri_c_test_smallint: (null)
FAIL: (null): cheri_c_test_smallint
[...]
TEST: cheri_c_test_intcapmath: (null)
FAIL: (null): cheri_c_test_intcapmath
[...]
SUMMARY: passed 158 failed 8 (1 expected)
```

**Comment by staceyson** [comment]
Running the cheri-c-tests cross built I see the following failures:

```
[...]
Running intcap
Assertion failed: (one == 1), function (null), file main, line 0.
Abort trap
FAILED: intcap
[...]
Running smallint
Assertion failed: (ptr == str), function (null), file main, line 0.
Abort trap
FAILED: smallint
Running stack_cap
In-address space security exception
FAILED: stack_cap
[...]
Running va_args
Assertion failed: (__builtin_memcap_length_get(ap) == argpairs * sizeof(void*) * 2), function (null), file printstuff, line 0.
Abort trap
FAILED: va_args
[...]
```

Also, the union test hangs in a failure loop (and had to remove it from run.sh).

**Comment by davidchisnall** [comment]
Thanks.  The last one is the most informative.  Do you think that you could stick in a debug `printf` before the assertion and see what the size of `ap` actually is (and, ideally, what `argpairs` is)?  It looks as if there might be something explicitly encoding the size of the capability when setting up the `va_list`.

**Comment by staceyson** [comment]
Here are each of the values in the assertion:

```
root@cheri:/var/tmp # ./va_args
__builtin_memcap_length_get(ap) (496) == argpairs (8) * sizeof(void*) (16) * 2
Assertion failed: (__builtin_memcap_length_get(ap) == argpairs * sizeof(void*) * 2), function (null), file printstuff, line 0.
```

**Comment by davidchisnall** [comment]
I believe this is fixed in CTSRD-CHERI/llvm@fa6c302adfb1912bdd92f18d0676bf51adf601b8.  At least, now the bounds are being set to 256 and it looks as if the arguments are being put in the right place.  The code for reading the arguments also looks correct (add 8 to get the next integer, round down to 16 and add 16 to  get the next capability), but please open another bug if it isn't.
